src/AppInstallerCLICore/CheckpointManager.cpp

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+#include "pch.h"
+#include "AppInstallerRuntime.h"
+#include "CheckpointManager.h"
+#include "Microsoft/CheckpointIndex.h"
+#include "Microsoft/SQLiteStorageBase.h"
+#include <Argument.h>
+#include <Command.h>
+
+namespace AppInstaller::CLI::Checkpoint
+{
+    void CheckpointManager::Initialize(GUID checkpointId)
+    {
+        // Initialize should only be called once.
+        THROW_HR_IF(E_UNEXPECTED, m_checkpointId != GUID_NULL);
+
+        bool createCheckpointIndex = (checkpointId == GUID_NULL);
+        if (createCheckpointIndex)
+        {
+            std::ignore = CoCreateGuid(&m_checkpointId);
+            AICLI_LOG(CLI, Info, << "Creating checkpoint index with guid: " << m_checkpointId);
+        }
+        else
+        {
+            m_checkpointId = checkpointId;
+            AICLI_LOG(CLI, Info, << "Opening checkpoint index with guid: " << m_checkpointId);
+        }
+
+        auto openDisposition = AppInstaller::Repository::Microsoft::SQLiteStorageBase::OpenDisposition::ReadWrite;
+        auto checkpointIndex = AppInstaller::Repository::Microsoft::CheckpointIndex::OpenOrCreateDefault(m_checkpointId, openDisposition);
+        if (!checkpointIndex)
+        {
+            AICLI_LOG(CLI, Error, << "Unable to open checkpoint index.");
+            THROW_HR_MSG(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), "The checkpoint state could not be found.");
+        }
+
+        m_checkpointIndex = std::move(checkpointIndex);
+
+        if (createCheckpointIndex)
+        {
+            m_checkpointIndex->SetClientVersion(AppInstaller::Runtime::GetClientVersion());
+        }
+    }
+
+    void CheckpointManager::Checkpoint(Execution::Context& context, Execution::CheckpointFlag checkpointFlag)
+    {
+        Execution::CheckpointFlag contextCheckpointFlag = context.GetTargetCheckpoint();
+
+        // If the context target checkpoint is ahead or equal to the current checkpoint, we have previously executed this state, load checkpoint state from index.
+        // If the context target checkpoint is behind the current checkpoint, we have not yet processed this state, save checkpoint state to index.
+        if (contextCheckpointFlag > checkpointFlag || contextCheckpointFlag == checkpointFlag)
+        {
+            LoadCheckpoint(context, checkpointFlag);
```

Comment:
This makes me think that you are attempting to design this where the command is not cooperating in the resume, but is simply calling a checkpoint function every so often.  From our discussions on this, I thought we agreed that that design was not going to work.